subsys/bluetooth/host/conn.c

```diff
@@ -890,7 +890,9 @@ void bt_conn_set_state(struct bt_conn *conn, bt_conn_state_t state)
 			tx_notify(conn);
 
 			/* Cancel Connection Update if it is pending */
-			if (conn->type == BT_CONN_TYPE_LE) {
+			if ((conn->type == BT_CONN_TYPE_LE) &&
+			    (k_work_delayable_busy_get(&conn->deferred_work) &
+			     (K_WORK_QUEUED | K_WORK_DELAYED))) {
 				k_work_cancel_delayable(&conn->deferred_work);
 			}
 
```

Comment:
The disconnect procedure is really convoluted, so I'm not sure about anything. But it seems to me that we must cancel `deferred_work` here (before any other thread can see that `conn` is `BT_CONN_DISCONNECTED`). Otherwise, we enter a race condition, where `deferred_work` can be executed before `conn_cleanup(conn)`. `deferred_work` will `bt_conn_unref(conn)`, so `conn_cleanup(conn)` get's a bad `conn` reference.

I think `k_work_cancel_sync()` may work here.

Reply:
no, use `k_work_cancel_sync` will cause deadlock

Reply:
since sysworkq has sended cmd and waiting for response, but if use `k_work_cancel_sync` will block BT RX, deadlock will occour.

Reply:
@alwa-nordic Ping

Reply:
@alwa-nordic I don't think the conn reference will be bad, because [the caller](https://github.com/zephyrproject-rtos/zephyr/blob/536fe6a948c9a546bb89dc73ab4adbd3b13b3e71/subsys/bluetooth/host/conn.c#L708) is just iterating on `acl_conns`, so most of them will actually already be disconnected. Since the stack is using coop priorities, then `conn_prepare_events` will not be pre-empted by the workqueue, and will just return `-ENOTCONN` when [the connection doesn't have any references](https://github.com/zephyrproject-rtos/zephyr/blob/536fe6a948c9a546bb89dc73ab4adbd3b13b3e71/subsys/bluetooth/host/conn.c#L671).

So I think this change should be okay, but just to have a bit more confidence, I will try to apply it downstream and run some robustness/stress tests.